<a href="https://colab.research.google.com/github/kxitika/Machine-Learning-Projects/blob/main/Credit%20Card%20Fraud%20Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CREDIT CARD FRAUD DETECTION

 **ABOUT THE DATASET**



Features V1, V2, … V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-sensitive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

**Acknowledgements**

The dataset has been collected and analysed during a research collaboration of Worldline and the Machine Learning Group (http://mlg.ulb.ac.be) of ULB (Université Libre de Bruxelles) on big data mining and fraud detection.

In [2]:
#importing the essential libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [4]:
#loading the dataset to a pandas dataframe
df = pd.read_csv('/content/creditcard.csv')

In [5]:
#printing the first 5 rows
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [6]:
#finding the statistics of the data
df.describe()

Time            V1            V2            V3            V4  \
count  23858.000000  23858.000000  23858.000000  23858.000000  23858.000000   
mean   18213.370609     -0.239141      0.198892      0.727022      0.248619   
std    11377.032190      1.894219      1.533073      1.724887      1.440938   
min        0.000000    -30.552380    -40.978852    -31.103685     -5.172595   
25%     6624.750000     -0.959528     -0.376134      0.287941     -0.658457   
50%    20564.000000     -0.288644      0.192491      0.874426      0.216440   
75%    29010.250000      1.164867      0.843146      1.505467      1.122367   
max    32954.000000      1.960497     16.713389      4.101716     11.927512   

                 V5            V6            V7            V8            V9  \
count  23858.000000  23858.000000  23858.000000  23858.000000  23858.000000   
mean      -0.188428      0.082789     -0.132080      0.022901      0.499391   
std        1.439894      1.328159      1.334691      1.378382      1.274199   
min      -42.147898    -23.496714    -26.548144    -41.484823     -7.175097   
25%       -0.767634     -0.663707     -0.595277     -0.167688     -0.330826   
50%       -0.218348     -0.179887     -0.066356      0.027183      0.422417   
75%        0.325281      0.474103      0.450460      0.286111      1.261519   
max       34.099309     22.529298     36.677268     20.007208     10.392889   

       ...           V21           V22           V23           V24  \
count  ...  23858.000000  23857.000000  23857.000000  23857.000000   
mean   ...     -0.041192     -0.137287     -0.038435      0.013626   
std    ...      0.837207      0.644720      0.526298      0.588573   
min    ...    -20.262054     -8.593642    -26.751119     -2.836627   
25%    ...     -0.253906     -0.561223     -0.173814     -0.325615   
50%    ...     -0.104848     -0.110305     -0.048570      0.064300   
75%    ...      0.066305      0.273088      0.074189      0.400535   
max    ...     22.614889      5.805795     13.876221      3.695503   

                V25           V26           V27           V28        Amount  \
count  23857.000000  23857.000000  23857.000000  23857.000000  23857.000000   
mean       0.127921      0.027504      0.013233      0.004932     73.976583   
std        0.435896      0.520213      0.401059      0.267790    212.761278   
min       -7.495741     -1.338556     -8.567638     -9.617915      0.000000   
25%       -0.134460     -0.337167     -0.066242     -0.009446      5.990000   
50%        0.164769     -0.046463      0.006465      0.020002     18.040000   
75%        0.407827      0.314884      0.094960      0.077392     66.000000   
max        5.525093      3.517346     11.135740      4.860769   7879.420000   

              Class  
count  23857.000000  
mean       0.003689  
std        0.060623  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000  

[8 rows x 31 columns]

0 -> legit transaction

1 -> fraudulent transaction


In [9]:
# getting the number of rows (examples) and columns(feartues)
df.shape

(23858, 31)

In [11]:
#getting datatype of each feature
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23858 entries, 0 to 23857
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    23858 non-null  int64  
 1   V1      23858 non-null  float64
 2   V2      23858 non-null  float64
 3   V3      23858 non-null  float64
 4   V4      23858 non-null  float64
 5   V5      23858 non-null  float64
 6   V6      23858 non-null  float64
 7   V7      23858 non-null  float64
 8   V8      23858 non-null  float64
 9   V9      23858 non-null  float64
 10  V10     23858 non-null  float64
 11  V11     23858 non-null  float64
 12  V12     23858 non-null  float64
 13  V13     23858 non-null  float64
 14  V14     23858 non-null  float64
 15  V15     23858 non-null  float64
 16  V16     23858 non-null  float64
 17  V17     23858 non-null  float64
 18  V18     23858 non-null  float64
 19  V19     23858 non-null  float64
 20  V20     23858 non-null  float64
 21  V21     23858 non-null  float64
 22

In [12]:
#checking the number of missing values in each feature
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [13]:
df[df.isnull().any(axis = 1)]

Time       V1        V2        V3        V4        V5        V6  \
23857  32954 -0.26422  1.107046  0.985544  0.346904  1.121035  0.069275   

             V7        V8       V9  ...       V21  V22  V23  V24  V25  V26  \
23857  1.307409 -0.681733  0.34665  ... -0.159083  NaN  NaN  NaN  NaN  NaN   

       V27  V28  Amount  Class  
23857  NaN  NaN     NaN    NaN  

[1 rows x 31 columns]

this means that all the missing values lie in a single example. So, a reasonable way to handle this would be to remove this row

In [17]:
#deleting the row with missing values
df = df.dropna()

In [18]:
df.shape

(23857, 31)

In [21]:
df['Class'].value_counts()

Class
0.0    23769
1.0       88
Name: count, dtype: int64

The above is an example of an imbalanced dataset, i.e, a dataset with an unequal class distribution. If not handled correctly, the model will always predict the majority class and still get a high accuracy which is misleading

In [ ]:
#separating the data for analysis

In [26]:
legit = df[df['Class'] == 0]
fraud = df[df['Class'] == 1]

In [27]:
print(legit.shape)
print(fraud.shape)

(23769, 31)
(88, 31)


In [28]:
#performing statistical measures on the data
legit['Amount'].describe()

count    23769.000000
mean        73.880199
std        212.541174
min          0.000000
25%          6.000000
50%         18.110000
75%         65.850000
max       7879.420000
Name: Amount, dtype: float64

In [29]:
fraud['Amount'].describe()

count      88.000000
mean      100.010000
std       265.845031
min         0.000000
25%         1.000000
50%         1.000000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [30]:
#comparing these values
df.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                   
0.0    18213.77782 -0.208135  0.175983   0.774952  0.226463 -0.166866   
1.0    17935.87500 -8.613716  6.376169 -12.221731  6.231847 -6.027247   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.092304 -0.101868  0.006906  0.512306  ...  0.038311 -0.043337   
1.0   -2.487080 -8.308784  4.351326 -2.987199  ...  0.714069  0.539387   

            V22       V23       V24       V25      V26       V27       V28  \
Class                                                                        
0.0   -0.136382 -0.037279  0.014613  0.127111  0.02694  0.010111  0.004578   
1.0   -0.381823 -0.350615 -0.252970  0.346695  0.17976  0.856336  0.100578   

           Amount  
Class              
0.0     73.880199  
1.0    100.010000  

[2 rows x 30 columns]

Using undersampling to handle imbalanced dataset by building a similar distribution of both the transactions

Undersampling works by reducing the number of samples from majority class so that both classes have a similar number of data points


In [33]:
legit_sample = legit.sample(n = 88)

In [34]:
#concatenating the 2 dataframes

In [36]:
new_df = pd.concat([legit_sample, fraud],axis = 0)

In [39]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
9178   13123  1.050793 -0.308407  0.024695  0.319205 -0.344702 -0.611736   
9102   12807 -0.646424  0.617808  1.348266 -1.510966 -0.926790  0.452395   
22579  32319 -0.343192 -0.528002  1.571971 -3.095768 -1.416688 -0.921063   
16040  27474 -0.803687  1.256307  1.220718 -0.107090 -0.315815 -0.608603   
14355  25463  1.239169  0.278347  0.655836  0.694930 -0.618880 -1.175139   

             V7        V8        V9  ...       V21       V22       V23  \
9178   0.109609 -0.218624  1.698267  ... -0.153383 -0.296553 -0.133183   
9102  -1.605291 -3.124127  0.250583  ...  3.337739 -0.071343  0.022288   
22579  0.082102 -0.037211  0.561505  ... -0.179927  0.105880  0.050578   
16040  0.287417  0.487778 -0.874606  ... -0.126648 -0.467246  0.031858   
14355  0.083091 -0.239202 -0.011539  ... -0.233889 -0.665190  0.174970   

            V24       V25       V26       V27       V28  Amount  Class  
9178   0.106681  0.381146  1.103681 -0.132261 -0.000012  115.00    0.0  
9102   0.168707  1.108434 -0.091833  0.220143  0.187315   15.00    0.0  
22579  0.577976  0.024491 -0.778491  0.152463  0.076290   69.98    0.0  
16040  0.317342 -0.186100  0.055045  0.136076  0.049182    9.83    0.0  
14355  0.712567  0.181814  0.073663 -0.026978  0.022664    0.89    0.0  

[5 rows x 31 columns]

In [38]:
new_df.shape

(176, 31)

In [40]:
new_df['Class'].value_counts()

Class
0.0    88
1.0    88
Name: count, dtype: int64

In [51]:
new_df.groupby('Class').mean()

Time        V1        V2         V3        V4        V5        V6  \
Class                                                                           
0.0    16583.875 -0.429820  0.225735   0.648464  0.243561 -0.059923 -0.014829   
1.0    17935.875 -8.613716  6.376169 -12.221731  6.231847 -6.027247 -2.487080   

             V7        V8        V9  ...       V20       V21       V22  \
Class                                ...                                 
0.0   -0.211829 -0.205051  0.603071  ...  0.011217  0.044110 -0.110553   
1.0   -8.308784  4.351326 -2.987199  ...  0.714069  0.539387 -0.381823   

            V23       V24       V25       V26       V27       V28      Amount  
Class                                                                          
0.0    0.037144 -0.001656  0.110416 -0.011668  0.070062  0.004832   61.868977  
1.0   -0.350615 -0.252970  0.346695  0.179760  0.856336  0.100578  100.010000  

[2 rows x 30 columns]

In [52]:
# splitting the data into features and target
x = new_df.drop('Class', axis = 1)

In [53]:
y = new_df['Class']

In [54]:
print(x)

        Time        V1        V2        V3        V4        V5        V6  \
9178   13123  1.050793 -0.308407  0.024695  0.319205 -0.344702 -0.611736   
9102   12807 -0.646424  0.617808  1.348266 -1.510966 -0.926790  0.452395   
22579  32319 -0.343192 -0.528002  1.571971 -3.095768 -1.416688 -0.921063   
16040  27474 -0.803687  1.256307  1.220718 -0.107090 -0.315815 -0.608603   
14355  25463  1.239169  0.278347  0.655836  0.694930 -0.618880 -1.175139   
...      ...       ...       ...       ...       ...       ...       ...   
18773  29753  0.269614  3.549755 -5.810353  5.809370  1.538808 -2.269219   
18809  29785  0.923764  0.344048 -2.880004  1.721680 -3.019565 -0.639736   
20198  30852 -2.830984  0.885657  1.199930  2.861292  0.321669  0.289966   
23308  32686  0.287953  1.728735 -1.652173  3.813544 -1.090927 -0.984745   
23422  32745 -2.179135  0.020218 -2.182733  2.572046 -3.663733  0.081568   

             V7        V8        V9  ...       V20       V21       V22  \
9178   0.1096

In [55]:
print(y)

9178     0.0
9102     0.0
22579    0.0
16040    0.0
14355    0.0
        ... 
18773    1.0
18809    1.0
20198    1.0
23308    1.0
23422    1.0
Name: Class, Length: 176, dtype: float64


In [56]:
#splitting the data into training and test data

In [58]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3,stratify = y)

In [59]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(123, 30)
(53, 30)
(123,)
(53,)


**TRAINING THE MODEL USING LOGISTIC REGRESSION**

In [62]:
model = LogisticRegression()
model.fit(x_train,y_train)


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [66]:
predicted_training = model.predict(x_train)

In [64]:
#model evaluation

In [70]:
print("Accuracy on training data is", accuracy_score(y_train,predicted_training) )

Accuracy on training data is 0.983739837398374


In [68]:
predicted_testing = model.predict(x_test)

In [71]:
print("Accuracy on testing data is ", accuracy_score(y_test,predicted_testing))

Accuracy on testing data is  0.9433962264150944


In this project, we developed a machine learning model to detect fraudulent credit card transactions using a real-world dataset. The goal was to build a classifier that can accurately distinguish between genuine and fraudulent transactions, which is a critical task in the financial industry to prevent monetary losses and ensure customer trust.


After preprocessing the data and training the model, we evaluated its performance:


 **Accuracy on Training Data: 98.37%**

 **Accuracy on Testing Data: 94.34%**

These results indicate that our model performs well on both the training and unseen test data, showing strong generalization capability with minimal overfitting.